devservices/utils/docker.py

```
@@ -19,9 +24,64 @@ def check_docker_daemon_running() -> None:
+    """Ensures all containers are healthy."""
+    with concurrent.futures.ThreadPoolExecutor() as healthcheck_executor:
+        futures = [
+            healthcheck_executor.submit(wait_for_healthy, container, status)
+            for container in containers
+        ]
+        for future in concurrent.futures.as_completed(futures):
+            future.result()
+
+
+def wait_for_healthy(container_name: str, status: Status) -> None:
+    """
+    Polls a Docker container's health status until it becomes healthy or a timeout is reached.
+    """
+    start = time.time()
+    while time.time() - start < HEALTHCHECK_TIMEOUT:
+        # Run docker inspect to get the container's health status
+        try:
+            # For containers with no healthchecks, the output will be "unknown"
+            result = subprocess.check_output(
+                [
+                    "docker",
+                    "inspect",
+                    "-f",
+                    "{{if .State.Health}}{{.State.Health.Status}}{{else}}unknown{{end}}",
+                    container_name,
+                ],
+                stderr=subprocess.DEVNULL,
+                text=True,
+            ).strip()
+        except subprocess.CalledProcessError as e:
+            raise DockerError(
+                command=f"docker inspect -f '{{.State.Health.Status}}' {container_name}",
+                returncode=e.returncode,
+                stdout=e.stdout,
+                stderr=e.stderr,
+            ) from e
+
+        if result == "healthy":
+            return
+        elif result == "unknown":
+            status.warning(
+                f"WARNING: Container {container_name} does not have a healthcheck"
+            )
+            return
+
+        # If not healthy, wait and try again
+        time.sleep(HEALTHCHECK_INTERVAL)
+
+    status.failure(
+        f"Container {container_name} did not become healthy within {HEALTHCHECK_TIMEOUT} seconds."
+    )
```

Instead of using status here, could we throw an error that we can catch higher up and write a failure status up there?